CDCgov/prime-reportstream
# Implement condition based routing and pruning

This PR implements code to condition mapping

Test Steps:
1. Run test suites
2. Submit the a report using the below `curl` command and `organizations.yml` file
   1. Use the following [organizations.yml](https://github.com/CDCgov/prime-reportstream/files/13815737/organizations.txt) (rename `.txt` to `.yml`)
   2. Starting with a clean database and reloaded settings/tables, submit a report using the `development.dev-elims` sender 
        ```zsh
        curl --location 'http://localhost:7071/api/reports' --header 'client: development.dev-elims' --header 'Content-Type: application/fhir+ndjson' --data-binary @src/test/resources/fhirengine/engine/routing/valid.fhir
        ```
   3. Open the azure storage explorer and browse to `batch/development.DEV_ELIMS_FHIR/`
      1. Download and view the most recent blob
      2. Verify none of these codes are present in the output
         - `95418-0`
         - `95417-2`
         - `95421-4`
         - `95419-8`
         - `AOE`
      3. Verify a `condition-code` extension is present with code `840539006` on observation w/ coding `94558-4`
    5. Browse to `batch/development.DEV_ELIMS_FHIR_FILTERED/`
       1. There should be no files here (or the directory may not exist)

## Changes
- Add bundle helpers for retrieving mapped conditions and filtering observations based on a list of codes
- Modify `FHIRRouter` to evaluate `mappedConditionFilter` against `Bundle` when finding receivers to route to
- Modify `FHIRRouter` to prune individual `Observation`s from a `Bundle` per receiver based on its `mappedConditionFilter`

## Checklist

### Testing
- [x] Added and modified tests?

## Linked Issues
- Fixes #12279 

## To Be Done
- #12705
- Remove orphaned resources (with corresponding test)
